### ISSUE TITLE
Clarification on typeguard vs pytypes

### ISSUE THREAD

**Issue by IvoMerchiers** [agronholm/typeguard#120]
On reading through issue #56, I noticed that @agronholm mentioned:

> My goal is to actually deprecate typeguard in favor of [pytypes](https://github.com/Stewori/pytypes) at some point.

I definitely understand where he is coming from, and focusing development efforts on one open-source project seems like the sensible thing to do.

However, that comment was from January 2019 and much might have changed in the meantime. Especially considering that pytypes seems to be less actively developed atm.

For new users (such as myself), it's hard to tell which package has certain advantages or to which one should contribute.

So it might be beneficial to have a section in the docs containing:
- If still relevant, any decision similar to 'typeguard is planned for deprecation in favor of pytypes', so that new contributors know on what they should focus.
- A short comparison between the two. Or at least a reference acknowledging its existence, so that users can read the docs for themselves.

**Comment by Stewori** [comment]
I can give the comparison to some extend:
typeguard is more focused on typechecking and more compact, has a smaller and cleaner codebase, is better tested and has a more up to date support of typing. It supports Python 3.5-3.8 AFAIK. (Used to support 3.3 and 3.4 until typing support was dropped in backports.)
It has no support for stubfiles, type comments and generics. On the other hand it is faster (for typechecking).

pytypes aims for a wider feature set, including support for stubfiles, type comments and generics. It also supports Python 2.7 and Jython via [type comments](https://www.python.org/dev/peps/pep-0484/#suggested-syntax-for-python-2-7-and-straddling-code). It provides a `get_type_hints` version that is aware of type comments and stubfiles. In general, it is intended as a typing toolbox rather than a pure runtime typechecker.
A main concern is to provide an exhaustive support of [`is_subtype`](https://github.com/Stewori/pytypes#is_subtypesubtype-supertype-bound_genericnone) (which still needs improvement). This is hard to get right for generics. It also has [`deep_type`](https://github.com/Stewori/pytypes#deep_typeobj-depthnone-max_samplenone), which works like PEP 484 aware `type(blah)`. `@typechecked` is mainly a proof of concept of this tooling, which is why it is slower than typeguard: In pytypes, the type of each object is really determined as a type object and then checked via `is_subtype`. An advantage is that it tells you precisely expected type vs. received type. Another reason why pytypes is slower is that it scans the source for type comments. This can be disabled to gain some speed.
It features a type-checking [`@overrides`](https://github.com/Stewori/pytypes#overriding-methods-in-type-safe-manner) annotation and automatically typechecks against parent method's types in OOP fashion if a child method is not typed. Nice thing here is that override checking can be done at class definition time rather than at (later "actual") runtime. I think the OOP integration is better than in typeguard, but I am not an expert of typeguard features. It also gets inner classes right.
Further, it has a set of utility methods that compensate regressions between typing releases and also typing bugs, e.g. [get_orig_class](https://github.com/Stewori/pytypes#get_orig_classobj-default_to__class__false). This goal has some overlap with [typing inspect](https://github.com/ilevkivskyi/typing_inspect), because pytypes predates typing inspect. In contrast to typing inspect, pytypes supports older typing releases and also gets Python 2.7 right. It also predates [pyannotate](https://github.com/dropbox/pyannotate) and can write stubfiles from runtime observations, e.g. during tests. However I did not much test and develop this feature since the raise of pyannotate.
Given the wider feature set, pytypes' codebase is far more complex than that of typeguard and harder to sustain. The changes in typing with Python 3.7 broke a lot and since then I haven't been able to work on anything else than fixing the breakage. Now Python 3.7 and 3.8 support is almost achieved but [not completely](https://github.com/Stewori/pytypes/issues/40).

I think neither project is going to deprecate. AFAIK @agronholm considered it but found pytypes codebase too difficult to work with. I admit it requires some cleanup but all my effort is bound by Python 3.7 support. Newer typing features are not (yet) supported for the same reason, e.g. literal types, async io, named tuples and others. ATM I cannot (afford to) work on pytypes but I expect to continue it at some point. Afterall it is developed without any funding. I don't know about typeguard's background.

Note that you can use other tools on top of pytypes via [`@annotations`](https://github.com/Stewori/pytypes#provide-info-from-type-comments-and-stubfiles-as-__annotations__-for-other-tools) which brings in type info from stubfiles and type comments as `__annotations__`. This way it should be possible to use pytypes just to bring in stubfile support and use typeguard on top for the typechecking if you prefer typeguard's way.

> to which one should contribute

I think both projects deserve it, but there are even more, e.g. [enforce](https://pypi.org/project/enforce/) or [pycontracts](https://andreacensi.github.io/contracts/) just to name two. It turns out, typing is a big enough field to allow for several approaches in parallel without any of them making the others obsolete. It is just too complex to have the one ultimate project (I know it because I intended pytypes as such). [This xkcd](https://xkcd.com/927/) pins it nicely. The "ultimate" project is too vulnerable against all the breakages. And the smaller, easier sustainable projects lack a bunch of features that are never to appear. E.g. AFAIK generics and stubfiles are *on principle* not going to get support in typeguard. Finally you have to decide which project is the best fit for your individual needs.

**Comment by agronholm** [comment]
I concur with @Stewori. I would have liked us to pool our efforts but turns out our goals are just too different.

**Comment by leycec** [comment]
[I'll just leave this here.](https://github.com/beartype/beartype#timings)

# tl;dr

**[Introducing beartype](https://github.com/beartype/beartype),** a [stupidly fast](https://github.com/beartype/beartype#timings) pure-Python runtime type-checker whose entire reason for existence is being... [stupidly fast](https://github.com/beartype/beartype#timings). The original use case for `beartype` was a [pure-Python multiphysics biology simulator implemented by my wife and I while working for Paul Allen](https://gitlab.com/betse/betse), but `beartype` slyly outgrew those humble beginnings sometime when I wasn't watching. Now it's it's own general-purpose rocket-fueled beast.

[`beartype`](https://github.com/beartype/beartype) is [less compatible than typeguard with most PEPs](https://github.com/beartype/beartype#features),<sup>that's bad</sup> but also dramatically faster. <sup>that's good!</sup> Most people who like typeguard will probably like [`beartype`](https://github.com/beartype/beartype), but some won't. So it goes.

I'd like to gratuitously thank typeguard for keeping the idolatrous flame of pure-Python runtime type checking alive all these long years. **You guys are amazing and continue to inspire.** Please don't go anywhere, because we're the last projects standing on this lonely summit.

I'd also to express my profound condolences to @RussBaz's [enforce](https://github.com/RussBaz/enforce) (which died a few years ago) and @Stewori's [pytypes](https://github.com/Stewori/pytypes) (which followed it last year). Which sucks, because now Python's runtime type-checking ecosystem has lost most of its biodiversity and is basically shambling shambolically on its last gangrenous legs. On the bright side, [`beartype`](https://github.com/beartype/beartype) is stupidly fast. Which rocks, because that was the last three months of my life. <sup>...*precious summer, come back to me!*</sup>

**Comment by agronholm** [comment]
Given that pytypes is unmaintained, this particular issue has become a moot point. As for beartype, I'm afraid I don't have the bandwidth to study it and compile a list of differences.

**Comment by leycec** [comment]
@agronholm: The differences are... *significant.* @beartype runs in constant-time (i.e., exhibits `O(1)` runtime complexity) – typically, in less than 10µs per type-check regardless of input size. `typeguard` runs in linear-time (i.e., exhibits `O(n)` runtime complexity).

@beartype will eventually support linear-time type-checking, too. Until then, any quantitative comparison is *really* apples-to-oranges. They serve different audiences. The world needs both.

Thanks so much for your continued maintenance of `typeguard`, @agronholm. :+1:

**Comment by agronholm** [comment]
> @agronholm: The differences are... _significant._ @beartype runs in constant-time (i.e., exhibits `O(1)` runtime complexity) – typically, in less than 10µs per type-check regardless of input size. `typeguard` runs in linear-time (i.e., exhibits `O(n)` runtime complexity).
> 
> @beartype will eventually support linear-time type-checking, too. Until then, any quantitative comparison is _really_ apples-to-oranges. They serve different audiences. The world needs both.
> 
> Thanks so much for your continued maintenance of `typeguard`, @agronholm. +1

Given that typeguard from v3.0.0 also checks only one item per collection by default, what are the remaining significant differences?
